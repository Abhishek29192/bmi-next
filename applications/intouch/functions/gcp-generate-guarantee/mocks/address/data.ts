import { Address } from "@bmi/intouch-api-types";
import { emptyNodes } from "../utils/graphileNodes";

const mockAddressBoilerplate = {
  createdAt: Date.now(),
  updatedAt: Date.now(),
  companiesByRegisteredAddressId: emptyNodes,
  companiesByTradingAddressId: emptyNodes,
  projectsByBuildingOwnerAddressId: emptyNodes,
  projectsBySiteAddressId: emptyNodes
};

export const mockCompanyAddress: Address = {
  id: 1,
  nodeId: "",
  firstLine: "23 Fairwell Street",
  secondLine: "Somewhere",
  town: "Fantatown",
  country: "Fantaland",
  postcode: "12300-002",
  ...mockAddressBoilerplate
};

export const mockSiteAddress: Address = {
  id: 2,
  nodeId: "",
  firstLine: "4719 Farwell Road",
  secondLine: "Tennyson",
  town: "Losino-Petrovskiy",
  country: "Russia",
  postcode: "19570-005",
  ...mockAddressBoilerplate
};

export const mockBuildingOwnerAddress: Address = {
  id: 3,
  nodeId: "",
  firstLine: "8 Boyd Terrace",
  secondLine: "Ohio",
  town: "Fl√≥rina",
  country: "Greece",
  postcode: "19570-035",
  ...mockAddressBoilerplate
};
