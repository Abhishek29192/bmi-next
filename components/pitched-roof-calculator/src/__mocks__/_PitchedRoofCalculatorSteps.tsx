import React from "react";

const PitchedRoofCalculatorSteps = () => <p>loaded</p>;

export default PitchedRoofCalculatorSteps;
