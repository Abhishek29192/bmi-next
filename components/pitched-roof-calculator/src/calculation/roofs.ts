import Roof7Illustration from "../images/roof-shapes/roof007.svg";
import Roof7DimensionsIllustration from "../images/roof-dimensions/7.svg";

const roof7 = {
  name: "Roof 7",
  type: "hipped",
  selectionIllustration: Roof7Illustration,
  dimensionsIllustration: Roof7DimensionsIllustration,
  dimensions: [
    {
      name: "A",
      type: "LENGTH"
    },
    {
      name: "B",
      type: "LENGTH"
    },
    {
      name: "P",
      type: "PITCH"
    }
  ],
  getMeasurements: (values) => {
    const A = parseFloat(values.A) * 100;
    const B = parseFloat(values.B) * 100;
    const P = parseFloat(values.P);
    const height = (A / 2) * Math.tan((P * Math.PI) / 180);
    const planeHeight = height / Math.sin((P * Math.PI) / 180);
    const planeTwoJointDisplacement = height / Math.tan((P * Math.PI) / 180);

    const hip = Math.sqrt((A / 2) ** 2 + planeHeight ** 2);
    const ridge = B - 2 * planeTwoJointDisplacement;
    return {
      faces: [
        {
          vertices: [
            { x: 0, y: 0 },
            {
              x: A / 2,
              y: planeHeight
            },
            { x: A, y: 0 }
          ],
          pitch: P
        },
        {
          vertices: [
            { x: 0, y: 0 },
            {
              x: A / 2,
              y: planeHeight
            },
            { x: A, y: 0 }
          ],
          pitch: P
        },
        {
          vertices: [
            { x: 0, y: 0 },
            {
              x: planeTwoJointDisplacement,
              y: planeHeight
            },
            {
              x: B - planeTwoJointDisplacement,
              y: planeHeight
            },
            { x: B, y: 0 }
          ],
          pitch: P
        },
        {
          vertices: [
            { x: 0, y: 0 },
            {
              x: planeTwoJointDisplacement,
              y: planeHeight
            },
            {
              x: B - planeTwoJointDisplacement,
              y: planeHeight
            },
            { x: B, y: 0 }
          ],
          pitch: P
        }
      ],
      lines: {
        hip: [hip, hip, hip, hip],
        ridge: [ridge],
        eave: [A, A, B, B],
        verge: []
      }
    };
  }
};

export default [roof7];
