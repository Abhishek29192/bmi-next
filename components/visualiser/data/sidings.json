{
  "sidings": [
    {
      "userId": 1,
      "name": "Svart",
      "previewRef": "",
      "diffuseMapRef": "public:8.jpg",
      "normalMapRef": "",
      "metallicRoughnessMapRef": "",
      "revision": 1,
      "createdUtc": "2021-01-19T17:00:52",
      "editedUtc": "2021-01-26T16:07:27",
      "id": 1,
      "revisionId": null,
      "isDraft": false,
      "creatorUser": null,
      "type": "Siding"
    },
    {
      "userId": 1,
      "name": "Mørk Brun",
      "previewRef": "",
      "diffuseMapRef": "public:9.jpg",
      "normalMapRef": "",
      "metallicRoughnessMapRef": "",
      "revision": 1,
      "createdUtc": "2021-01-19T17:00:59",
      "editedUtc": "2021-01-26T16:07:37",
      "id": 2,
      "revisionId": null,
      "isDraft": false,
      "creatorUser": null,
      "type": "Siding"
    },
    {
      "userId": 1,
      "name": "Lys Brun",
      "previewRef": "",
      "diffuseMapRef": "public:10.jpg",
      "normalMapRef": "",
      "metallicRoughnessMapRef": "",
      "revision": 1,
      "createdUtc": "2021-01-19T17:01:04",
      "editedUtc": "2021-01-26T16:07:45",
      "id": 3,
      "revisionId": null,
      "isDraft": false,
      "creatorUser": null,
      "type": "Siding"
    },
    {
      "userId": 1,
      "name": "Grå",
      "previewRef": "",
      "diffuseMapRef": "public:11.jpg",
      "normalMapRef": "",
      "metallicRoughnessMapRef": "",
      "revision": 1,
      "createdUtc": "2021-01-20T12:06:14",
      "editedUtc": "2021-01-26T16:07:52",
      "id": 4,
      "revisionId": null,
      "isDraft": false,
      "creatorUser": null,
      "type": "Siding"
    },
    {
      "userId": 1,
      "name": "Rød",
      "previewRef": "",
      "diffuseMapRef": "public:12.jpg",
      "normalMapRef": "",
      "metallicRoughnessMapRef": "",
      "revision": 1,
      "createdUtc": "2021-01-20T12:06:25",
      "editedUtc": "2021-01-26T16:07:58",
      "id": 5,
      "revisionId": null,
      "isDraft": false,
      "creatorUser": null,
      "type": "Siding"
    },
    {
      "userId": 1,
      "name": "Hvit",
      "previewRef": "",
      "diffuseMapRef": "public:13.jpg",
      "normalMapRef": "",
      "metallicRoughnessMapRef": "",
      "revision": 1,
      "createdUtc": "2021-01-20T12:06:39",
      "editedUtc": "2021-01-26T16:08:06",
      "id": 6,
      "revisionId": null,
      "isDraft": false,
      "creatorUser": null,
      "type": "Siding"
    },
    {
      "userId": 1,
      "name": "Murstein",
      "previewRef": "public:210.jpg",
      "diffuseMapRef": "public:210.jpg",
      "normalMapRef": "public:208.jpg",
      "metallicRoughnessMapRef": "public:211.jpg",
      "revision": 1,
      "createdUtc": "2021-01-25T11:06:35",
      "editedUtc": "2021-01-26T16:08:16",
      "id": 7,
      "revisionId": null,
      "isDraft": false,
      "creatorUser": null,
      "type": "Siding"
    }
  ]
}
