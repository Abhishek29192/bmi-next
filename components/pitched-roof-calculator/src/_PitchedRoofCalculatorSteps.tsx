import React, { useContext, useEffect, useMemo, useState } from "react";
import CalculatorStepper from "@bmi/calculator-stepper";
import { BMI as brandLogo } from "@bmi/logo";
import Icon from "@bmi/icon";
import { AnalyticsContext } from "./helpers/analytics";
import { getMicroCopy, MicroCopyContext } from "./helpers/microCopy";
import RoofSelection from "./_RoofSelection";
import RoofDimensions from "./_RoofDimensions";
import TileSelection from "./_TileSelection";
import TileOptions, { TileOptionsSeletions } from "./_TileOptions";
import VariantSelection from "./_VariantSelection";
import UnderlaySelection from "./_UnderlaySelection";
import Guttering, { GutteringSelections } from "./_Guttering";
import { calculateArea } from "./calculation/calculate";
import Results from "./_Results";
import { EmailFormValues } from "./types/EmailFormValues";
import protrusionTypes from "./calculation/protrusions";
import { DimensionsValues, Measurements, Roof } from "./types/roof";
import styles from "./_PitchedRoofCalculatorSteps.module.scss";
import { CONTINGENCY_PERCENTAGE_TEXT } from "./calculation/constants";
import { Data, MainTile, MainTileVariant, Underlay } from "./types";

export type Step =
  | "select-roof"
  | "enter-dimensions"
  | "select-tile"
  | "select-variant"
  | "tile-options"
  | "select-underlay"
  | "guttering"
  | "your-solution-contains";

export type PitchedRoofCalculatorStepsProps = {
  data: Data;
  isDebugging?: boolean;
  selected: Step;
  setSelected: (value: Step) => void;
  sendEmailAddress: (values: EmailFormValues) => Promise<void>;
};

const PitchedRoofCalculatorSteps = ({
  data, // TODO: use here
  isDebugging,
  selected,
  setSelected,
  sendEmailAddress
}: PitchedRoofCalculatorStepsProps) => {
  const copy = useContext(MicroCopyContext);
  const pushEvent = useContext(AnalyticsContext);

  const [roof, setRoof] = useState<Roof | undefined>(undefined);
  const [dimensions, setDimensions] = useState<DimensionsValues>({});
  const [tile, setTile] = useState<MainTile | undefined>(undefined);
  const [variant, setVariant] =
    useState<MainTileVariant | undefined>(undefined);
  const [tileOptions, setTileOptions] =
    useState<TileOptionsSeletions | undefined>(undefined);
  const [underlay, setUnderlay] = useState<Underlay | undefined>(undefined);
  const [guttering, setGuttering] =
    useState<GutteringSelections | undefined>(undefined);

  const selectRoof = (newRoof: Roof) => {
    setSelected("enter-dimensions");
    if (newRoof === roof) return;
    setRoof(newRoof);
    setDimensions({});
  };

  const saveDimensions = (e: React.FormEvent<Element>, values: object) => {
    e.preventDefault();

    pushEvent({
      event: "dxb.button_click",
      id: "rc-dimensions",
      label: getMicroCopy(copy, "roofDimensions.nextLabel"),
      action: "selected"
    });

    // Nice to have: check if the dimensions are different before resetting
    setDimensions(values as DimensionsValues);
    setTile(undefined);
    setSelected("select-tile");
  };

  const selectTile = (newTile: MainTile) => {
    setSelected("select-variant");
    if (newTile === tile) return;
    setTile(newTile);
    setVariant(undefined);
  };

  const selectVariant = (newVariant: MainTileVariant) => {
    setSelected("tile-options");
    if (newVariant === variant) return;
    setVariant(newVariant);
    setTileOptions(undefined);
  };

  const saveAndMove = <T extends any>(
    e: React.FormEvent<Element>,
    values: T,
    set: React.Dispatch<React.SetStateAction<T | undefined>>,
    next: Step
  ) => {
    e.preventDefault();
    set(values);
    setSelected(next);
  };

  const measurements: Measurements | null = useMemo(() => {
    if (!roof || !dimensions) return null;
    let { faces, lines } = roof.getMeasurements(dimensions);

    if (Array.isArray(dimensions.protrusions)) {
      for (const { type, ...protrusionDimensions } of dimensions.protrusions) {
        // eslint-disable-next-line security/detect-object-injection
        if (protrusionTypes[type]?.getMeasurements) {
          // eslint-disable-next-line security/detect-object-injection
          const { faces: pFaces, lines: pLines } = protrusionTypes[
            type
          ].getMeasurements({
            ...protrusionDimensions,
            roofPitch: dimensions[roof.roofPitchField] // Says on what face of the roof the protrusion is
          });
          faces.push(...pFaces);
          Object.keys(lines).forEach((line) =>
            // eslint-disable-next-line security/detect-object-injection
            lines[line].push(...pLines[line])
          );
        }
      }
    }

    return {
      faces,
      lines,
      area: calculateArea(faces)
    };
  }, [roof, dimensions]);

  const formattedArea = ((measurements?.area || 0) / 10000).toFixed(2);

  useEffect(() => {
    if (selected === "your-solution-contains") {
      pushEvent({
        event: "dxb.button_click",
        id: "rc-solution-m2",
        label: "mÂ²",
        action: formattedArea
      });
    }
  }, [selected, formattedArea]);

  return (
    <div className={styles["PitchedRoofCalculatorSteps"]}>
      <Icon source={brandLogo} className={styles["logo"]} />
      <CalculatorStepper selected={selected}>
        <CalculatorStepper.Step
          key="select-roof"
          title={getMicroCopy(copy, "roofSelection.title")}
          subtitle={getMicroCopy(copy, "roofSelection.subtitle")}
        >
          <RoofSelection select={selectRoof} selected={roof} />
        </CalculatorStepper.Step>
        <CalculatorStepper.Step
          key="enter-dimensions"
          title={getMicroCopy(copy, "roofDimensions.title")}
          subtitle={getMicroCopy(copy, "roofDimensions.subtitle")}
          nextLabel={getMicroCopy(copy, "roofDimensions.nextLabel")}
          nextButtonOnClick={saveDimensions}
          backLabel={getMicroCopy(copy, "roofDimensions.backLabel")}
          backButtonOnClick={() => {
            pushEvent({
              event: "dxb.button_click",
              id: "rc-dimensions",
              label: getMicroCopy(copy, "roofDimensions.backLabel"),
              action: "selected"
            });
            setSelected("select-roof");
          }}
        >
          {roof ? <RoofDimensions roof={roof} dimensions={dimensions} /> : null}
        </CalculatorStepper.Step>
        <CalculatorStepper.Step
          key="select-tile"
          title={getMicroCopy(copy, "tileSelection.title")}
          subtitle={getMicroCopy(copy, "tileSelection.subtitle")}
          backLabel={getMicroCopy(copy, "tileSelection.backLabel")}
          backButtonOnClick={() => {
            pushEvent({
              event: "dxb.button_click",
              id: "rc-select-tile",
              label: getMicroCopy(copy, "tileSelection.backLabel"),
              action: "selected"
            });
            setSelected("enter-dimensions");
          }}
        >
          <TileSelection
            tiles={data.mainTiles}
            select={selectTile}
            selected={tile}
            dimensions={dimensions}
          />
        </CalculatorStepper.Step>
        <CalculatorStepper.Step
          key="select-variant"
          title={getMicroCopy(copy, "variantSelection.title")}
          subtitle={getMicroCopy(copy, "variantSelection.subtitle")}
          backLabel={getMicroCopy(copy, "variantSelection.backLabel")}
          backButtonOnClick={() => {
            pushEvent({
              event: "dxb.button_click",
              id: "rc-select-tile-colour",
              label: getMicroCopy(copy, "variantSelection.backLabel"),
              action: "selected"
            });
            setSelected("select-tile");
          }}
        >
          {tile ? (
            <VariantSelection
              select={selectVariant}
              selected={variant}
              dimensions={dimensions}
              tile={tile}
            />
          ) : null}
        </CalculatorStepper.Step>
        <CalculatorStepper.Step
          key="tile-options"
          title={getMicroCopy(copy, "tileOptions.title")}
          subtitle={getMicroCopy(copy, "tileOptions.subtitle")}
          nextLabel={getMicroCopy(copy, "tileOptions.nextLabel")}
          nextButtonOnClick={(e, values) => {
            pushEvent({
              event: "dxb.button_click",
              id: "rc-options-accessories",
              label: getMicroCopy(copy, "tileOptions.nextLabel"),
              action: "selected"
            });
            saveAndMove(e, values, setTileOptions, "select-underlay");
          }}
          backLabel={getMicroCopy(copy, "tileOptions.backLabel")}
          backButtonOnClick={() => {
            pushEvent({
              event: "dxb.button_click",
              id: "rc-options-accessories",
              label: getMicroCopy(copy, "tileOptions.backLabel"),
              action: "selected"
            });
            setSelected("select-variant");
          }}
        >
          {variant && tileOptions ? (
            <TileOptions variant={variant} selections={tileOptions} />
          ) : null}
        </CalculatorStepper.Step>
        <CalculatorStepper.Step
          key="select-underlay"
          title={getMicroCopy(copy, "underlaySelection.title")}
          subtitle={getMicroCopy(copy, "underlaySelection.subtitle")}
          nextLabel={getMicroCopy(copy, "underlaySelection.nextLabel")}
          nextButtonOnClick={(e, values) => {
            pushEvent({
              event: "dxb.button_click",
              id: "rc-select-underlay",
              label: getMicroCopy(copy, "underlaySelection.nextLabel"),
              action: "selected"
            });
            saveAndMove(e, values as Underlay, setUnderlay, "guttering");
          }}
          backLabel={getMicroCopy(copy, "underlaySelection.backLabel")}
          backButtonOnClick={() => {
            pushEvent({
              event: "dxb.button_click",
              id: "rc-select-underlay",
              label: getMicroCopy(copy, "underlaySelection.backLabel"),
              action: "selected"
            });
            setSelected("tile-options");
          }}
        >
          <UnderlaySelection
            options={data.underlays}
            selected={underlay}
            dimensions={dimensions}
          />
        </CalculatorStepper.Step>
        <CalculatorStepper.Step
          key="guttering"
          title={getMicroCopy(copy, "guttering.title")}
          subtitle={getMicroCopy(copy, "guttering.subtitle")}
          nextLabel={getMicroCopy(copy, "guttering.nextLabel")}
          nextButtonOnClick={(e, values) => {
            pushEvent({
              event: "dxb.button_click",
              id: "rc-select-guttering",
              label: getMicroCopy(copy, "guttering.nextLabel"),
              action: "selected"
            });
            saveAndMove(e, values, setGuttering, "your-solution-contains");
          }}
          backLabel={getMicroCopy(copy, "guttering.backLabel")}
          backButtonOnClick={() => {
            pushEvent({
              event: "dxb.button_click",
              id: "rc-select-guttering",
              label: getMicroCopy(copy, "guttering.backLabel"),
              action: "selected"
            });
            setSelected("select-underlay");
          }}
          linkLabel={getMicroCopy(copy, "guttering.skipLabel")}
          linkOnClick={() => {
            pushEvent({
              event: "dxb.button_click",
              id: "rc-select-guttering",
              label: getMicroCopy(copy, "guttering.skipLabel"),
              action: "selected"
            });
            setSelected("your-solution-contains");
          }}
        >
          {guttering ? (
            <Guttering
              selections={guttering}
              gutters={data.gutters}
              gutterHooks={data.gutterHooks}
            />
          ) : null}
        </CalculatorStepper.Step>
        <CalculatorStepper.Step
          isForm={false}
          key="your-solution-contains"
          title={getMicroCopy(copy, "results.title")}
          subtitle={getMicroCopy(copy, "results.subtitle", {
            contingency: CONTINGENCY_PERCENTAGE_TEXT
          })}
          paragraph={
            <span>
              {`${getMicroCopy(copy, "results.areaLabel")}: ${formattedArea}m`}
              <sup>2</sup>
            </span>
          }
          backLabel={getMicroCopy(copy, "results.backLabel")}
          backButtonOnClick={() => {
            pushEvent({
              event: "dxb.button_click",
              id: "rc-solution",
              label: getMicroCopy(copy, "results.backLabel"),
              action: "selected"
            });
            setSelected("guttering");
          }}
          linkLabel={getMicroCopy(copy, "results.startOverLabel")}
          linkOnClick={() => {
            pushEvent({
              event: "dxb.button_click",
              id: "rc-solution",
              label: getMicroCopy(copy, "results.startOverLabel"),
              action: "selected"
            });
            setSelected("select-roof");
            setRoof(undefined);
          }}
        >
          {measurements && variant && tileOptions && underlay && guttering && (
            <Results
              underlays={data.underlays}
              gutters={data.gutters}
              gutterHooks={data.gutterHooks}
              {...{
                isDebugging,
                measurements,
                variant,
                tileOptions,
                underlay,
                guttering,
                sendEmailAddress
              }}
            />
          )}
        </CalculatorStepper.Step>
      </CalculatorStepper>
    </div>
  );
};

export default PitchedRoofCalculatorSteps;
