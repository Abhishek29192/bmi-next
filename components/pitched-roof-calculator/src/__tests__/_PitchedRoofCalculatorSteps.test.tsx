import roofs from "../calculation/roofs";
import data from "../samples/data.json";
import en from "../samples/copy/en.json";

const components = [
  "_RoofSelection",
  "_RoofDimensions",
  "_TileSelection",
  "_TileOptions",
  "_VariantSelection",
  "_UnderlaySelection",
  "_Guttering",
  "_Results"
];

let componentProps = {};

let React;
let render;
let act;
let MicroCopy;
let PitchedRoofCalculatorSteps;

beforeEach(() => {
  jest.resetModules();
  jest.clearAllMocks();
  componentProps = {};
  React = require("react");
  render = require("@testing-library/react").render;
  act = require("@testing-library/react").act;
  MicroCopy = require("../helpers/microCopy").MicroCopy;
  components.forEach((component) => {
    jest.mock(`../${component}`, () => ({
      __esModule: true,
      default: jest.fn().mockImplementation((props) => {
        componentProps[component] = props;
        return <p>Rendering {component}</p>;
      })
    }));
  });
});

describe("PitchedRoofCalculatorSteps component", () => {
  it("renders correctly", () => {
    PitchedRoofCalculatorSteps =
      require("../_PitchedRoofCalculatorSteps").default;

    const { container } = render(
      <MicroCopy.Provider values={en}>
        <PitchedRoofCalculatorSteps
          selected="select-roof"
          setSelected={jest.fn()}
          sendEmailAddress={jest.fn()}
          data={data as any}
        />
      </MicroCopy.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it("moves through the steps", async () => {
    let renderedStep;
    const stepProps = {};
    jest.mock(`@bmi/calculator-stepper`, () => {
      const CalculatorStepper = ({ selected, children }) => {
        renderedStep = selected;
        return React.Children.toArray(children)
          .filter((item) => (item.key + "").substr(2) === selected)
          .map((element) =>
            React.cloneElement(element, {
              passedKey: (element.key + "").substr(2)
            })
          );
      };

      CalculatorStepper.Step = jest.fn().mockImplementation((props) => {
        stepProps[props.passedKey] = props;
        return (
          <div>
            <p>Step {props.passedKey}</p>
            {props.children}
          </div>
        );
      });

      return {
        __esModule: true,
        default: CalculatorStepper
      };
    });

    PitchedRoofCalculatorSteps =
      require("../_PitchedRoofCalculatorSteps").default;

    let selected = "select-roof";
    const setSelected = jest
      .fn()
      .mockImplementation((selection) => (selected = selection));
    const sendEmailAddress = jest.fn();

    const getComponent = (selected) => (
      <MicroCopy.Provider values={en}>
        <PitchedRoofCalculatorSteps
          selected={selected}
          setSelected={setSelected}
          sendEmailAddress={sendEmailAddress}
          data={data as any}
        />
      </MicroCopy.Provider>
    );

    const { rerender } = render(getComponent(selected));

    expect(renderedStep).toBe("select-roof");

    act(() => componentProps["_RoofSelection"].select(roofs[0]));
    rerender(getComponent(selected));
    expect(renderedStep).toBe("enter-dimensions");

    act(() =>
      stepProps["enter-dimensions"].nextButtonOnClick(
        { preventDefault: () => {} },
        {
          A: "9",
          B: "15",
          P: "22",
          protrusions: []
        }
      )
    );
    rerender(getComponent(selected));
    expect(renderedStep).toBe("select-tile");

    const tile = data.mainTiles[0];

    act(() => componentProps["_TileSelection"].select(tile));
    rerender(getComponent(selected));
    expect(renderedStep).toBe("select-variant");

    const variant = tile.variants[0];

    act(() => componentProps["_VariantSelection"].select(variant));
    rerender(getComponent(selected));
    expect(renderedStep).toBe("tile-options");

    act(() =>
      stepProps["tile-options"].nextButtonOnClick(
        { preventDefault: () => {} },
        {
          verge: "none",
          ridge: variant.ridgeOptions[0].externalProductCode,
          ventilation: "none"
        }
      )
    );
    rerender(getComponent(selected));
    expect(renderedStep).toBe("select-underlay");

    const underlay = data.underlays[0];

    act(() =>
      stepProps["select-underlay"].nextButtonOnClick(
        { preventDefault: () => {} },
        {
          underlay: underlay.externalProductCode
        }
      )
    );
    rerender(getComponent(selected));
    expect(renderedStep).toBe("guttering");

    act(() => stepProps["guttering"].linkOnClick());
    rerender(getComponent(selected));
    expect(renderedStep).toBe("your-solution-contains");

    act(() => stepProps["your-solution-contains"].backButtonOnClick());
    rerender(getComponent(selected));
    expect(renderedStep).toBe("guttering");

    const gutter = data.gutters[0];
    const gutterVariant = gutter.variants[0];
    const gutterHook = data.gutterHooks[0];

    act(() =>
      stepProps["guttering"].nextButtonOnClick(
        { preventDefault: () => {} },
        {
          guttering: gutter.name,
          gutteringVariant: gutterVariant.externalProductCode,
          gutteringHook: gutterHook.externalProductCode,
          downPipes: 10,
          downPipeConnectors: 15
        }
      )
    );
    rerender(getComponent(selected));
    expect(renderedStep).toBe("your-solution-contains");

    [
      ["your-solution-contains", "guttering"],
      ["guttering", "select-underlay"],
      ["select-underlay", "tile-options"],
      ["tile-options", "select-variant"],
      ["select-variant", "select-tile"],
      ["select-tile", "enter-dimensions"],
      ["enter-dimensions", "select-roof"]
    ].forEach(([current, previous]) => {
      act(() => stepProps[current].backButtonOnClick());
      rerender(getComponent(selected));
      expect(renderedStep).toBe(previous);
    });
  });
});
