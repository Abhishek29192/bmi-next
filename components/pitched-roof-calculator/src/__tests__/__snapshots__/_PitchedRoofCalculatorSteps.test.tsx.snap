// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PitchedRoofCalculatorSteps component renders correctly 1`] = `
<div>
  <div
    class="PitchedRoofCalculatorSteps"
  >
    <svg
      class="logo"
    />
    <div
      class="CalculatorStepper"
    >
      <form
        class="Form"
      >
        <h4
          class="MuiTypography-root step-title Typography MuiTypography-h4"
        >
          Choose your roof type
        </h4>
        <h6
          class="MuiTypography-root computation Typography MuiTypography-h6"
        />
        <p
          class="MuiTypography-root subtitle Typography MuiTypography-body1"
        >
          Get started by letting us know the roof shape that most closely resembles your roof
        </p>
        <div
          class="content"
        >
          <p>
            Rendering 
            _RoofSelection
          </p>
        </div>
        <hr
          class="MuiDivider-root hr"
        />
      </form>
    </div>
  </div>
</div>
`;
